controller/token_storage_blackbox_test.go

```diff
@@ -0,0 +1,249 @@
+		panic("invalid test " + err.Error()) // bug
+	}
+
+	referrerClientURL := "https://localhost.example.ui/home"
+	req.Header.Add("referer", referrerClientURL)
+
+	prms := url.Values{
+		"for": {"https://github.com/sbose78"},
+	}
+
+	goaCtx := goa.NewContext(service.Context, rw, req, prms)
+	tokenCtx, err := app.NewRetrieveTokenContext(goaCtx, req, goa.New("TokenService"))
+	require.Nil(rest.T(), err)
+
+	err = controller.Retrieve(tokenCtx)
+	require.NotNil(rest.T(), err)
+	expectedHeaderValue := "LINK url=https://auth.localhost.io/api/link, description=\"github token is missing. Link github account\""
+	assert.Contains(rest.T(), rw.Header().Get("WWW-Authenticate"), expectedHeaderValue)
+}
+
+// Not present in keycloak and identity is not the system.
+func (rest *TestTokenStorageREST) TestRetrieveExternalTokenIdentityNotPresent() {
+	resource.Require(rest.T(), resource.Database)
+	identity := testsupport.TestIdentity // using an Identity which has no existence the database.
+	rest.mockKeycloakExternalTokenServiceClient.scenario = "unlinked"
+	service, controller := rest.SecuredControllerWithIdentity(identity)
+	test.RetrieveTokenUnauthorized(rest.T(), service.Context, service, controller, "https://github.com/a/b")
+}
+
+// Not present in keycloak but present in DB.
+func (rest *TestTokenStorageREST) TestRetrieveExternalTokenPresentInDB() {
+	resource.Require(rest.T(), resource.Database)
+	identity, err := testsupport.CreateTestIdentity(rest.DB, uuid.NewV4().String(), "KC")
+	require.Nil(rest.T(), err)
+
+	rest.mockKeycloakExternalTokenServiceClient.scenario = "unlinked"
+	service, controller := rest.SecuredControllerWithIdentity(identity)
+
+	r := &goa.RequestData{
+		Request: &http.Request{Host: "api.example.org"},
+	}
+	providerConfig, err := rest.providerConfigFactory.NewOauthProvider(context.Background(), r, "https://github.com/a/b")
+	require.Nil(rest.T(), err)
+
+	expectedToken := provider.ExternalToken{
+		ProviderID: providerConfig.ID(),
+		Scope:      providerConfig.Scopes(),
+		IdentityID: identity.ID,
+		Token:      "1234-from-db",
+	}
+	rest.externalTokenRepository.Create(context.Background(), &expectedToken)
+
+	// This call should end up in a failed KC response , but a positive retrieval from the database.
+	_, tokenResponse := test.RetrieveTokenOK(rest.T(), service.Context, service, controller, "https://github.com/a/b")
+	require.Equal(rest.T(), expectedToken.Token, tokenResponse.AccessToken)
+	require.Equal(rest.T(), expectedToken.Scope, tokenResponse.Scope)
+	require.Equal(rest.T(), "bearer", tokenResponse.TokenType)
+}
+
+func (rest *TestTokenStorageREST) TestRetrieveExternalTokenBadRequest() {
+	resource.Require(rest.T(), resource.Database)
+	identity := testsupport.TestIdentity
+	service, controller := rest.SecuredControllerWithIdentity(identity)
+	test.RetrieveTokenBadRequest(rest.T(), service.Context, service, controller, "")
+}
+
+func (rest *TestTokenStorageREST) createRandomUserAndIdentityForStorage() account.Identity {
```

Comment:
Just remove this method. Use testsupport.CreateTestIdentity(rest.DB, uuid.NewV4().String(), "KC") instead everywhere.

Reply:
Forgot to remove this method, I've already switched over  https://github.com/fabric8-services/fabric8-auth/pull/140/files/0633d70379b9d5fe87bbfb834b0614c83ac0cf45#diff-ad1b5dfda94882afaabdd2e1a239388fR157 as per your feedback !

Reply:
Thanks , done. there were a couple of places I had missed. https://github.com/fabric8-services/fabric8-auth/pull/140/commits/256ed7924cd7602fecc891e23b4f382b37231603